### PR TITLE
feature/firewall-sso-tenant-aware

### LINKED ISSUE #6368: Firewall SSO: add support for tenants
**Is your feature request related to a problem? Please describe.**
In a multi-tenant configuration, there is no way to dedicate a firewall SSO configuration for a specific tenant.

You can play on network to do SSO or on roles to trigger updates for specific tenant but they have to be unique which is not always the case (for networks).

**Describe the solution you'd like**
Dedicate firewall SSO per tenant.

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Make Firewall SSO Tenant Aware

Impacts
=======
Firewall SSO.

NEWS file entries
=======================
Enhancements
-------------
* Firewall SSO is tenant aware.

Delete branch after merge
=========================
NO

Issue
=========================
Fixes #6368

Checklist
=========
- [ ] Document the feature
- [x] Add unit tests
- [ ] Add acceptance tests (TestLink)

### CODE REVIEW

File: addons/upgrade/to-11.0-firewall_sso-conf.pl [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/perl
+
+=head1 NAME
+
+to-10.2-pfdetect-conf -
```
Comment (fdurand) [inline addons/upgrade/to-11.0-firewall_sso-conf.pl]: not pfdetect

File: addons/upgrade/to-11.0-firewall_sso-conf.pl [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+=head1 DESCRIPTION
+
+Add the new field tenant_id to the detect parsers
```
Comment (fdurand) [inline addons/upgrade/to-11.0-firewall_sso-conf.pl]: wrong description

File: go/caddy/pfsso/pfsso.go [hunk 3]
```diff
@@ -79,19 +78,30 @@ func buildPfssoHandler(ctx context.Context) (*PfssoHandler, error) {
...
 		msg := fmt.Sprintf("Error while decoding payload: %s", err)
 		log.LoggerWContext(ctx).Error(msg)
 		return nil, 0, errors.New(msg)
 	}
 
+	tenant_id := r.Header.Get("X-PacketFence-Tenant-Id")
+	if tenant_id == "" {
+		tenant_id = defaultTenantID
+	} else if _, err := strconv.ParseInt(tenant_id, 10, 32); err != nil {
+		log.LoggerWContext(ctx).Warn(fmt.Sprintf("Can't parse X-PacketFence-Tenant-Id '%s' into an int (%s). Will not specify timeout for request.", tenant_id, err))
```
Comment (fdurand) [inline go/caddy/pfsso/pfsso.go]: why timeout when we deal with tenant ?

File: addons/upgrade/to-11.0-firewall_sso-conf.pl [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+=head1 DESCRIPTION
+
+Add the new field tenant_id to Firewall SSO config
+
+=cut
+
+use strict;
+use warnings;
+use lib qw(/usr/local/pf/lib);
```
Comment (fdurand) [inline addons/upgrade/to-11.0-firewall_sso-conf.pl]: can you add /usr/local/pf/lib_perl/lib/perl5 too

File: go/firewallsso/base.go [hunk 5]
```diff
@@ -216,6 +218,16 @@ func (fw *FirewallSSO) MatchesNetwork(ctx context.Context, info map[string]strin
 	return false
 }
 
+func (fw *FirewallSSO) MatchesTenant(ctx context.Context, info map[string]string) bool {
+	tenant_id, found := info["tenant_id"]
+	if !found {
+		tenant_id = "1"
```
Comment (fdurand) [inline go/firewallsso/base.go]: Use a constant
const defaultTenantID = "1"

**Review by fdurand — COMMENTED** [review]
Just 2 details.
@satkunas is it ok for the admin gui part ?

**Comment by satkunas** [comment]
@fdurand ui looks good

**Comment by fdurand** [comment]
OK so let's @jrouzierinverse fix the 2 details and we will be ready to merge

**Comment by jrouzierinverse** [comment]
Done

**Pull request merged**